### PR TITLE
Generalize yield statement function type

### LINKED ISSUE #1011: Need to add Generator support
PEP 484 defines a Generator class, a subclass of Iterator that also specifies the types of send() and the return value (for `yield from`). I'm not 100% sure that this doesn't require changes to mypy.

### PR DESCRIPTION
**Pull request by sakana**
Functions containing yield statements can now return any supertype
of generator. This fixes #1011.

In making this pull request, I noticed that the allowed return types of functions containing `yield`s are different depending on whether or not the `yield` is a statement, expression, or a `yield from`.  I'll make a PR to make them all behave consistently next.

### CODE REVIEW

File: mypy/messages.py [hunk 1]
```diff
@@ -30,7 +30,7 @@
 INVALID_EXCEPTION = 'Exception must be derived from BaseException'
 INVALID_EXCEPTION_TYPE = 'Exception type must be derived from BaseException'
 INVALID_RETURN_TYPE_FOR_YIELD = \
-    'Iterator function return type expected for "yield"'
+    'Generator or supertype function return type expected for "yield"'
```
Comment (gvanrossum) [inline mypy/messages.py]: This sentence sounds a bit odd. Maybe we can throw it around, e.g.
```
The return type of a generator function should be "Generator" or a supertype thereof.
```
(The mention of "yield" in the original message is kind of a distraction -- the problem is not with the yield statement/expression but with the function's type declaration. Extra points if you can find the line number of the 'def' line.)

**Comment by gvanrossum** [comment]
LGTM except for the phrasing of the error message.

I'm looking forward to seeing the other issue fixed too!

File: mypy/messages.py (hunk 1 above)
Comment (sakana) [inline mypy/messages.py]: I agree -- that rephrasing is much better.

Comment (sakana) [inline mypy/messages.py]: I don't see any clear way of setting the error message line to the function, though...

File: mypy/messages.py [hunk 2]
```diff
@@ -30,7 +30,7 @@
 INVALID_EXCEPTION = 'Exception must be derived from BaseException'
 INVALID_EXCEPTION_TYPE = 'Exception type must be derived from BaseException'
 INVALID_RETURN_TYPE_FOR_YIELD = \
-    'Iterator function return type expected for "yield"'
+    'The return type of a generator function should be "Generator" or a supertype thereof'
```
Comment (kirbyfan64) [inline mypy/messages.py]: What about "or one of its supertypes"?

Comment (sakana) [inline mypy/messages.py]: I think that's even clearer.  Thanks!

**Pull request merged**